### PR TITLE
LSTM without layers

### PR DESCRIPTION
**Pull request by ahtsan**
* Add `lstm`, `lstm_model`, `categorical_lstm_policy_with_model`.
* Unit tests for the above.
* Test TRPO with `categorical_lstm_policy_with_model`

### CODE REVIEW

**Comment by ryanjulian** [comment]
It seems you may have forgotten to cover the `state_include_action` case for LSTM:

https://codecov.io/gh/rlworkgroup/garage/commit/07d0d3155ce9961affb152589e9dc3545f487831#D12-207

File: garage/tf/core/layers.py [hunk 1]
```diff
@@ -49,12 +49,12 @@ def as_tuple(x, n, t=None):
         x = (x, ) * n
 
     if (t is not None) and not all(isinstance(v, t) for v in x):
-        raise TypeError("expected a single value or an iterable "
-                        "of {0}, got {1} instead".format(t.__name__, x))
+        raise TypeError('expected a single value or an iterable '
```
[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/core/layers.py]: It's done by the pre-push checks.

File: garage/tf/algos/npo.py [hunk 2]
```diff
@@ -390,7 +390,7 @@ def _build_entropy_term(self, i):
                 policy_dist_info = self.policy.dist_info_sym(
                     i.obs_var,
                     i.policy_state_info_vars,
-                    name='policy_dist_info')
+                    name='policy_dist_info_entropy')
```
[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/algos/npo.py]: I renamed it because we can't call `dist_info_sym` with the same name twice (because the underlying model will build twice with the same name, which is not allowed). I think Chang is going to fix it by changing the `name_scope` to `variable_scope`. @CatherineSue

File: garage/tf/policies/base2.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 """Policy base classes without Parameterized."""
 import tensorflow as tf
 
+from garage.misc.tensor_utils import flatten_tensors, unflatten_tensors
```
[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/policies/base2.py]: After discussing with Zhanpeng and ran some test, I think `tf.VariableScope` is working properly.

[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/policies/base2.py]: I will fix it in another PR since modifying `Policy2` will affect other policies as well.

File: garage/tf/algos/npo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Changing it from `name_scope` to `variable_scope` won't fix the error. The cause is in Model instead of here.

File: garage/tf/policies/categorical_lstm_policy_with_model.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            prev_action_var = tf.cast(prev_action_var, tf.float32)
+            all_input_var = tf.concat(
+                axis=2, values=[obs_var, prev_action_var])
+        else:
+            all_input_var = obs_var
+
+        with tf.variable_scope(self._variable_scope):
+            outputs, _, _, _, _, _ = self.model.build(
+                all_input_var,
+                *self.model.networks['default'].inputs[1:4],
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply garage/tf/policies/categorical_lstm_policy_with_model.py]: yeah this is pretty crazy. what even is `inputs[1:4]`?

[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/policies/categorical_lstm_policy_with_model.py]: It is crazy. I will fix it.

File: examples/tf/trpo_cartpole_recurrent_with_model.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            baseline=baseline,
+            max_path_length=100,
+            discount=0.99,
+            max_kl_step=0.01,
+            optimizer=ConjugateGradientOptimizer,
+            optimizer_args=dict(
+                hvp_approach=FiniteDifferenceHvp(base_eps=1e-5)))
+
+        runner.setup(algo, env)
+        runner.train(n_epochs=100, batch_size=4000)
```
Comment (ryanjulian) [inline examples/tf/trpo_cartpole_recurrent_with_model.py]: this is a lot of iters. how long does this test take?

File: garage/tf/algos/npo.py [hunk 6]
```diff
@@ -384,14 +387,9 @@ def _build_policy_loss(self, i):
 
             return loss, pol_mean_kl
 
-    def _build_entropy_term(self, i):
+    def _build_entropy_term(self, i, policy_dist_info):
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: please don't make these functions dependent on anything but `i`

File: garage/tf/models/base.py [hunk 7]
```diff
@@ -188,8 +188,7 @@ def build(self, *inputs, name=None):
         if not self._networks:
             # First time building the model, so self._networks are empty
             # We store the variable_scope to reenter later when we reuse it
-            with tf.variable_scope(
-                    self._name, reuse=False) as self._variable_scope:
+            with tf.variable_scope(self._name) as self._variable_scope:
```
Comment (ryanjulian) [inline garage/tf/models/base.py]: please explicitly assign the new a scope into a member variable -- do not capture it as a coincidence of a context manager

File: garage/tf/models/base.py [hunk 8]
```diff
@@ -188,8 +188,7 @@ def build(self, *inputs, name=None):
         if not self._networks:
```
Comment (ryanjulian) [inline garage/tf/models/base.py]: don't forget to update the types in the docstring

File: garage/tf/models/base.py [hunk 9]
```diff
@@ -210,20 +209,29 @@ def build(self, *inputs, name=None):
                     network = Network()
                     network._inputs = inputs
                     network._outputs = self._build(*inputs, name)
-        spec = self.network_output_spec()
-        if spec:
-            c = namedtuple(network_name,
-                           [*spec, 'input', 'output', 'inputs', 'outputs'])
+        in_spec = self.network_input_spec()
```
Comment (ryanjulian) [inline garage/tf/models/base.py]: this logic about `spec` is getting kind of complicated and i'm not sure i understand what's going on.

can you explain what the purpose of this code is? perhaps it would be good to break it into a well-documented helper function.

can you also help me understand the purpose of the input and output specs?

File: tests/helpers.py [hunk 10]
```diff
@@ -87,6 +87,77 @@ def convolve(_input, filter_weights, filter_bias, strides, filter_sizes,
...
+    w_ho = np.full((num_units, num_units), w_h_init)
+    b_o = np.full((num_units, ), b_init)
+
+    w_x_ifco = np.concatenate([w_xi, w_xf, w_xc, w_xo], axis=1)
+    w_h_ifco = np.concatenate([w_hi, w_hf, w_hc, w_ho], axis=1)
+
+    x_ifco = np.matmul(input_val, w_x_ifco)
+    h_ifco = np.matmul(step_hidden, w_h_ifco)
+
+    x_i, x_f, x_c, x_o = np.split(x_ifco, 4, axis=1)  # noqa: E501, pylint: disable=unbalanced-tuple-unpacking
```
Comment (ryanjulian) [inline tests/helpers.py]: what is the justifcation for disabling E501 and the pylint warning?

File: garage/tf/algos/npo.py (hunk 6 above)
Comment (ryanjulian) [reply garage/tf/algos/npo.py]: it is very easy to make a loss function which is a tangle of calls to other functions. we instead make sure all components of the loss function are only dependent on `i` (the input spec, basically) which makes it easy to follow.

this will be much easier once the loss function is a model.

Comment (ahtsan) [reply garage/tf/algos/npo.py]: I see. I modified that since `_build_entropy_term` actually reuses the same `policy_dist_info` symbolic tensor with `_build_policy_loss`. I will come up with another way.

File: garage/tf/models/base.py (hunk 7 above)
Comment (ahtsan) [reply garage/tf/models/base.py]: Forgot to modify it here. Will fix.

File: tests/helpers.py (hunk 10 above)
Comment (ahtsan) [reply tests/helpers.py]: Since I can't find out what's the reason pylint is complaining about that error, so I disable the pylint checking and since it is too long, I disabled E501 as well. I will dig into it and see what pylint is talking about.

File: garage/tf/models/base.py (hunk 9 above)
Comment (ahtsan) [reply garage/tf/models/base.py]: Basically the `specs` tells us what the inputs and outputs are. There are 4 cases:
### Case 1: Single input/output model
Then we don't need any specs, since we don't need extra attribute for the networks. We only need `input,inputs,output,outputs` which are the default ones in network.
### Case 2: Single input/Multiple output model
We set `network_output_spec()` so that we can have assign "names" to the outputs, e.g. 
```python
def network_output_spec(self):
    return ['state', 'action']
```
Then we can do `model.networks['default'].state` and `model.networks['default'].action`
In this case, we don't need `network_input_spec()` since there is only one input. What we need is only `input`, which are already included by default.
### Case 3: Multiple input/Single output model
Same as above, but instead of having `network_output_spec`, we have `network_input_spec` this time.
### Case 4: Multiple input/output model
We have both `network_output_spec` and `network_input_spec`. Then we can have sth like
`model.networks['default'].state` and `model.networks['default'].action` as the outputs and `model.networks['default'].state_input` and `model.networks['default'].action_input` as the inputs, to make it more readable.

File: tests/helpers.py (hunk 10 above)
Comment (ryanjulian) [reply tests/helpers.py]: perhaps is one of your return values itself a tuple or list?

File: garage/tf/models/base.py (hunk 9 above)
Comment (ryanjulian) [reply garage/tf/models/base.py]: this seems pretty complex -- is there a way we can always have an input/output spec, even for multi -input and -output models?

i think *always* having a spec would simplify the logic significantly.

Comment (ahtsan) [reply garage/tf/models/base.py]: Do you mean we make `specs` not optional?
Right now by default the specs are a empty list, which means networks will only have `input,inputs,output,outputs`.

File: examples/tf/trpo_cartpole_recurrent_with_model.py (hunk 5 above)
Comment (ahtsan) [reply examples/tf/trpo_cartpole_recurrent_with_model.py]: Settings are the same as the existing `trpo_cartpole_recurrent`.`trpo_cartpole_recurrent` takes  ~400s while `trpo_cartpole_recurrent_with_model` takes ~460s.

File: garage/tf/models/base.py (hunk 9 above)
Comment (ryanjulian) [reply garage/tf/models/base.py]: i mean that you internally maintain the `specs` data structure regardless of whether the user has chosen to do multi input/output or not. then the single-input and single-output cases don't need to be handled with so much special logic.

the outer interface can stay the same.

File: examples/tf/trpo_cartpole_recurrent_with_model.py (hunk 5 above)
Comment (ryanjulian) [reply examples/tf/trpo_cartpole_recurrent_with_model.py]: oops i thought this was a test. it is concerning that the model version takes 10% longer. i wonder why, but we don't need to worry much about it now.

File: garage/tf/algos/npo.py (hunk 6 above)
Comment (ryanjulian) [reply garage/tf/algos/npo.py]: you can just not reuse it...

Comment (ahtsan) [reply garage/tf/algos/npo.py]: We have to reuse it if we don't do dist info sym with a different name. Because each network name is unique. That's why I add "_entropy" after that.

Comment (ahtsan) [reply garage/tf/algos/npo.py]: We can just append a "_2" to the name in the second call of dist info sym.

Comment (ryanjulian) [reply garage/tf/algos/npo.py]: sure

File: tests/helpers.py (hunk 10 above)
Comment (ahtsan) [reply tests/helpers.py]: no... They are all `numpy.ndarray`.

Comment (ahtsan) [reply tests/helpers.py]: Could be because interpreter cannot ensure the right hand side can be split into 4. Is there any other way to pass pylint for this?

Comment (ryanjulian) [reply tests/helpers.py]: generally pylint rules try to encourage being very explicit and writing code so that it can be statically analyzed for errors (e.g. using pylint).

in this case i think pylint has good point, because the `np.split` operation is totally static (all inputs are known at compile time), but it has been written as a dynamic operation unnecessarily.

the rationale for this is that code is read much more often than it is written, so it's usually better to make it longer and easier to read rather than shorted and easier to write.

maybe it would be clearer if you just wrote out. this way a reader doesn't have to look up documentation for np.split to find out how your array is unpacked.

```python
x_i, x_f, x_c, x_o = x_ifco[:, :4], x_ifco[:, 4:8], x_ifco[:, 8:12], x_ifco[:, 12:16]
# or
x_i = x_ifco[:, :4]
x_f = x_ifco[:, 4:8]
x_c = x_ifco[:, 8:12]
x_o = x_ifco[:, 12:16]
```

Comment (ahtsan) [reply tests/helpers.py]: Got it.

File: tests/garage/tf/core/test_lstm.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
+import numpy as np
+import tensorflow as tf
+
+from garage.tf.core.lstm import lstm
+from tests.fixtures import TfGraphTestCase
+
+
+class TestLSTM(TfGraphTestCase):
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply tests/garage/tf/core/test_lstm.py]: i think overall this testing strategy is okay, but please add some simple checks too, e.g.

* output length vs input length
* gradient path (or lack thereof) between pairs of inputs/outputs
* non-zero hidden states
* etc.

perhaps you can take a look as the tests from lasagne for some ideas: https://github.com/Lasagne/Lasagne/blob/master/lasagne/tests/layers/test_recurrent.py

File: garage/tf/models/base.py (hunk 7 above)
Comment (ryanjulian) [reply garage/tf/models/base.py]: bump
